sentence_transformers/SentenceTransformer.py

```diff
@@ -87,6 +88,7 @@ def __init__(
         token: Optional[Union[bool, str]] = None,
         use_auth_token: Optional[Union[bool, str]] = None,
         truncate_dim: Optional[int] = None,
+        model_args: Optional[Dict[str, Any]] = None
     ):
```

Comment:
I'm not a fan of passing a dictionary of arguments. I think this can be rather confusing, as it's unclear what can be passed here. Instead, I think we should allow passing of specific, useful parameters, starting with:
* `torch_dtype`
* `attn_implementation`

Additionally, we can add `**model_kwargs` as the final parameter. So:
```suggestion
        torch_dtype: Optional[Union[str, torch.dtype]] = None,
        attn_implementation: Optional[str] = None,
        **model_kwargs,
    ):
        model_kwargs = {
            "torch_dtype": torch_dtype,
            "attn_implementation": attn_implementation,
            **model_kwargs,
        }
```

([docs](https://huggingface.co/docs/transformers/main_classes/model#transformers.PreTrainedModel.from_pretrained))

Then the user can use:
```python
model = SentenceTransformer("all-mpnet-base-v2", torch_dtype=torch.float16, low_cpu_mem_usage=True)
```
etc.

The explicit parameters (torch_dtype, attn_implementation) will then be clear to the user. The big question that remains is which parameters should be used, e.g. there are a lot of options, but many are not commonly used for most: https://huggingface.co/docs/transformers/main_classes/model#transformers.PreTrainedModel.from_pretrained.torch_dtype

Additionally, we might want to do some magic around `device_map` ([docs](https://huggingface.co/docs/transformers/main_classes/model#transformers.PreTrainedModel.from_pretrained.device_map)) and `device` that currently exists. Not really sure about the best approach for that yet.

Reply:
Thank you for taking the time to review this PR and look at the issue.


I've added torch_dtype and attn_implementation params to the signature

- I'm concerned if making all the other kwargs as model_kwargs will put us in a difficult situation if we ever need pass-through tokenizer_kwargs. Probably it is best to add params on need basis?
- For the documentation, I'm still directing to Transformers. Would you prefer directly phrasing the HF documentation? I think it would become an overhead to keep the two aligned as more features are added to HF (especially for attn_implementation)
- I'm also unclear on how to handle device_map. Probably it is best to defer model_kwargs for now to not end up in in a situation where both device_map and device are passed in.

Additionally, could you please suggest how would you want me to test for attn_implementation? Would a test that mocks the underlying call and validates correct arguments are passed to it suffice?

Reply:
> I'm concerned if making all the other kwargs as model_kwargs will put us in a difficult situation if we ever need pass-through tokenizer_kwargs. Probably it is best to add params on need basis?

This is a concern of mine too, also as we add more features, we could run into some unexpected changes. However, I think it might still be the best solution. However, I think the tokenizer kwargs are not that commonly used: most of the time any changes there result in degraded model performance.

> For the documentation, I'm still directing to Transformers. Would you prefer directly phrasing the HF documentation? I think it would become an overhead to keep the two aligned as more features are added to HF (especially for attn_implementation)

I think it makes sense to copy the latest version of HF for now, and I can change it when necessary. This will always be a bit tricky as someone can use Sentence Tranformers with different versions of Transformers and have different behaviour. There's not really any getting around that. 

I'm not sure about device_map either. I think it makes sense to just "leave it" for now.

As for the attn_implementation - yes, that would be more than sufficient. Actual tests would require additional complex installations, and currently only a few architectures support anything other than "eager". [I'm actively working on that](https://github.com/huggingface/transformers/pull/30450), though.

Reply:
@muellerzr I'm quite curious about your opinion on this topic: do you think we should introduce `**model_kwargs` to pass arguments to `AutoModel`? It could cause some unexpected behaviour in the future if we add new explicit keyword arguments (as they might not be directly passed down to `AutoModel` anymore), but it has the upside of a much cleaner API.

Reply:
Post talking with Amy & Arthur (as they handle more of that side):

It depends a bit on how likely `sentence-transformers` is to either 
1) want to modify the values 
2) ever take these values themselves. 

I can imagine sentence-transformers might want to take e.g. torch_dtype in the future. This means it can no longer be in `model_kwargs` , so you’d need to e.g. pop it out. What happens if the torch_dtype wanted by sentence transformers is now different e.g. acceptable values than what we can pass to `AutoModel`? The problem with passing all the kwargs blindly, is you don’t even know if there’s going to be argument clashes for a new kwarg!

Your version here is adaptable and flexible, very nice! You'll still have that future kwarg clashing, but here it's more pragmatic and explicit.

In general: kwargs are a pain, and if it were in raw transformers we'd say no probably, unless we were in your shoes and calling another lib we don't control, in which being more flexible and doing exactly this could be good.

(Paraphrasing Amy's response to me)

Reply:
Thank you for the suggestions and explanation, I've updated the PR.